src/sentry/auth/staff.py

```diff
@@ -0,0 +1,318 @@
+            logger.info(
+                "staff.session-expired",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        return data
+
+    def _populate(self, current_datetime=None):
+        if current_datetime is None:
+            current_datetime = timezone.now()
+
+        request = self.request
+        user = getattr(request, "user", None)
+        if not hasattr(request, "session"):
+            data = None
+        elif not (user and user.is_staff):
+            data = None
+        else:
+            data = self.get_session_data(current_datetime=current_datetime)
+
+        if not data:
+            self._set_logged_out()
+        else:
+            self._set_logged_in(expires=data["exp"], token=data["tok"], user=user)
+
+            if not self.is_active:
+                if self._inactive_reason:
+                    logger.warning(
+                        "staff.%s",
+                        self._inactive_reason,
+                        extra={
+                            "ip_address": request.META["REMOTE_ADDR"],
+                            "user_id": request.user.id,
+                        },
+                    )
+                else:
+                    logger.warning(
+                        "staff.inactive-unknown-reason",
+                        extra={
+                            "ip_address": request.META["REMOTE_ADDR"],
+                            "user_id": request.user.id,
+                        },
+                    )
+
+    def _set_logged_in(self, expires, token, user, current_datetime=None):
+        # we bind uid here, as if you change users in the same request
+        # we wouldn't want to still support staff auth (given
+        # the staff check happens right here)
+        assert user.is_staff
+        if current_datetime is None:
+            current_datetime = timezone.now()
+        self.token = token
+        self.uid = str(user.id)
+        # the absolute maximum age of this session
+        self.expires = expires
+        # do we have a valid staff session?
+        self.is_valid = True
+        # is the session active? (it could be valid, but inactive)
+        self._is_active, self._inactive_reason = self.is_privileged_request()
+        self.request.session[SESSION_KEY] = {
+            "exp": self.expires.strftime("%s"),
+            "idl": (current_datetime + IDLE_MAX_STAFF_SESSION_AGE).strftime("%s"),
+            "tok": self.token,
+            # XXX(dcramer): do we really need the uid safety mechanism
+            "uid": self.uid,
+        }
+
+    def _set_logged_out(self):
+        self.uid = None
+        self.expires = None
+        self.token = None
+        self._is_active = False
+        self._inactive_reason = None
+        self.is_valid = False
+        self.request.session.pop(SESSION_KEY, None)
+
+    def set_logged_in(self, user, current_datetime=None):
```

Comment:
note the difference between `set_logged_in` for `Staff` vs `Superuser`
- `Staff` does not exist for self-hosted, is not enabled automatically upon login, and does not require a form to enable (only SSO/2FA)
- Therefore `Staff` does not have a `_needs_validation` function
- Therefore `Staff` also does not have form checks